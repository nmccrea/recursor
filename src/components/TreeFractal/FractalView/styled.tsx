import styled from "styled-components"

const Viewport = styled.div`
  position: fixed;
  top: 0px;
  left: 0px;
  width: 100vw;
  height: 100vh;
  display: flex;
  justify-content: center;
  align-items: center;
  overflow: hidden;
`

export { Viewport }
