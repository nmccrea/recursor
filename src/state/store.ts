/**
 * Configures and exports the redux store and other top-level Redux resources.
 * @packageDocumentation
 */

import { configureStore, combineReducers } from "@reduxjs/toolkit"
import similarities from "./similarities/reducer"

const rootReducer = combineReducers({ similarities })
const createStore = () => configureStore({ reducer: rootReducer })

export { createStore }
export type RootState = ReturnType<typeof rootReducer>
export type Selector<ReturnType> = (state: RootState) => ReturnType | undefined
