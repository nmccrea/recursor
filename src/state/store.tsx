/**
 * Configures and exports the redux store and other top-level Redux resources.
 * @packageDocumentation
 */

import { combineReducers, configureStore } from "@reduxjs/toolkit"

const rootReducer = combineReducers({})
export type RootState = ReturnType<typeof rootReducer>
export default configureStore({ reducer: rootReducer })
