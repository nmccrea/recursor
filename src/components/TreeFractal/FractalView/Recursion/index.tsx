import React from "react"
import styled from "styled-components"
import { useSelector } from "react-redux"
import { selectAll } from "../../../../state/similarities/selectors"
import {
  Similarity,
  Translation,
  Scale,
  Angle,
  Depth,
  Color,
} from "../../../../state/similarities/types"

// Use `styled-components` transient props to prevent prop forwarding.
// See: https://styled-components.com/docs/api#transient-props
interface TransformProps {
  $translation: Translation
  $scale: Scale
  $angle: Angle
  $color: Color
}

const Geometry = styled.div<TransformProps>`
  position: absolute;
  top: 0px;
  bottom: 0px;
  left: 0px;
  right: 0px;

  background-color: ${({ $color }) => $color};

  transform-origin: center bottom;
  transform: ${({ $translation, $scale, $angle }) =>
    `translateY(-${
      100 * $translation
    }%) scale(${$scale}) rotate(${$angle}rad)`};
`

interface RecursionProps extends Similarity {
  currentDepth: Depth
}

const Recursion = ({
  id,
  depth,
  currentDepth,
  translation,
  scale,
  angle,
  color,
}: RecursionProps) => {
  if (currentDepth >= depth) return null

  const similarities = useSelector(selectAll)
  return (
    <Geometry
      $translation={translation}
      $scale={scale}
      $angle={angle}
      $color={color}
      className={`similarity-${id} depth-${currentDepth + 1}`}
    >
      {similarities.map((similarity) => (
        <Recursion
          {...similarity}
          currentDepth={currentDepth + 1}
          key={similarity.id}
        />
      ))}
    </Geometry>
  )
}

export default Recursion
