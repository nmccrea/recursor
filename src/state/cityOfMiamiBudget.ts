/**
 * TODO: type
 * TODO: test
 * TODO: doc
 */

import { RootState } from "./store"
import {
  createAsyncThunk,
  createSlice,
  CaseReducer,
  PayloadAction,
} from "@reduxjs/toolkit"

// State
enum Status {
  Idle = "idle",
  Pending = "pending",
  Fulfilled = "fulfilled",
}

interface CityOfMiamiBudgetState {
  status: Status
  data?: object
}

const INITIAL_STATE: CityOfMiamiBudgetState = { status: Status.Idle }

// Case Reducers
const fetchDataPending: CaseReducer<CityOfMiamiBudgetState, PayloadAction> = (
  previousState
) => ({
  ...previousState,
  status: Status.Pending,
})

const fetchDataFulfilled: CaseReducer<
  CityOfMiamiBudgetState,
  PayloadAction<object>
> = (previousState, action) => ({
  ...previousState,
  status: Status.Fulfilled,
  data: action.payload,
})

// Async Thunks
const fetchData = createAsyncThunk("fetchData", async (url: string) => {
  const response = await fetch(url)
  return await response.json()
})

// Slice
const { reducer } = createSlice({
  name: "cityOfMiamiBudget",
  initialState: INITIAL_STATE,
  reducers: {},
  extraReducers: {
    [fetchData.pending.type]: fetchDataPending,
    [fetchData.fulfilled.type]: fetchDataFulfilled,
  },
})

// Selectors
const getStatus = (state: RootState) =>
  (state.cityOfMiamiBudget && state.cityOfMiamiBudget.status) || undefined

const getData = (state: RootState) =>
  (state.cityOfMiamiBudget && state.cityOfMiamiBudget.data) || undefined

export default reducer
export { fetchData, getStatus, getData, Status }
