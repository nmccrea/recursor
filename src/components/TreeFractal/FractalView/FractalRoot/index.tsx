import React from "react"
import styled from "styled-components"
import Recursion from "../Recursion"

const RootGeometry = styled.div`
  width: 6px;
  height: 120px;
  position: relative;
  transform: translateY(100%);
`

const FractalRoot = () => (
  <RootGeometry>
    <Recursion
      id="root"
      color="black"
      translation={0}
      scale={1}
      angle={0}
      depth={0}
      currentDepth={0}
    />
  </RootGeometry>
)

export default FractalRoot
