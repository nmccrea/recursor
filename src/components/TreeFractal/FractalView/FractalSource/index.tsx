import React from "react"
import styled from "styled-components"
import Recursion from "../Recursion"

const SourceGeometry = styled.div`
  width: 6px;
  height: 120px;
  position: relative;
  transform: translateY(100%);
`

const FractalSource = () => (
  <SourceGeometry>
    <Recursion
      id="source"
      color="black"
      translation={0}
      scale={1}
      angle={0}
      depth={1}
      currentDepth={0}
    />
  </SourceGeometry>
)

export default FractalSource
