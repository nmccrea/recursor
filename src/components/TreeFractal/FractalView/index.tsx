import React from "react"
import FractalSource from "./FractalSource"

const FractalView = () => (
  <div
    style={{
      width: "95vw",
      height: "70vh",
      display: "flex",
      justifyContent: "center",
      alignItems: "flex-end",
    }}
  >
    <FractalSource />
  </div>
)

export default FractalView
