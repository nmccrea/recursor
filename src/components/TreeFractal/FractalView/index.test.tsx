import React from "react"
import { Provider } from "react-redux"
import { render } from "@testing-library/react"
import { createStore, RootState } from "../../../state/store"
import FractalView from "."

describe("`<FractalView />`", () => {
  it("renders correctly", () => {
    const state: RootState = {
      similarities: {
        ids: ["test/a", "test/b"],
        entities: {
          "test/a": {
            id: "test/a",
            translation: 0.994,
            scale: 0.543,
            angle: -3.012 * Math.PI,
            depth: 3,
            color: "green",
          },
          "test/b": {
            id: "test/b",
            translation: 1.042,
            scale: 0.452,
            angle: 1.902 * Math.PI,
            depth: 2,
            color: "red",
          },
        },
      },
    }

    const { container } = render(
      <Provider store={createStore(state)}>
        <FractalView />
      </Provider>
    )

    expect(container).toMatchSnapshot()
  })
})
