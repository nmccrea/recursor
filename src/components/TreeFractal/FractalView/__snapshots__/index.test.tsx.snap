// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`\`<FractalView />\` renders correctly 1`] = `
<div>
  <div
    class="sc-AxiKw bkSRTf"
  >
    <div
      class="sc-AxirZ eJFmNH"
    >
      <div
        class="sc-AxjAm lhlKPj similarity-source depth-1"
      >
        <div
          class="sc-AxjAm enjJjR similarity-test/a depth-2"
        >
          <div
            class="sc-AxjAm enjJjR similarity-test/a depth-3"
          />
        </div>
        <div
          class="sc-AxjAm hDRcvy similarity-test/b depth-2"
        >
          <div
            class="sc-AxjAm enjJjR similarity-test/a depth-3"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
